### PR TITLE
Implement updater on macOS

### PR DESCRIPTION
**Pull request by kthchew**
Following my idea to use a third-party framework for updates in #466, this implements the Sparkle updater framework for macOS. Makes work towards #60, but does not entirely fix it as this only works on macOS.

To release a new update, you add an entry to an RSS feed. Here is an example of such a feed, which I will temporarily use to test this PR:

```
<?xml version="1.0" standalone="yes"?>
<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" version="2.0">
    <channel>
        <title>PolyMC</title>
        <item>
            <title>1.2.1</title>
            <pubDate>Wed, 20 Apr 2022 23:45:13 -0400</pubDate>
            <sparkle:fullReleaseNotesLink>
                https://polymc.org/news/
            </sparkle:fullReleaseNotesLink>
            <sparkle:releaseNotesLink>
                https://raw.githubusercontent.com/PolyMC/polymc.github.io/master/src/news/release-1.2.1.md
            </sparkle:releaseNotesLink>
            <sparkle:version>1.2.1</sparkle:version>
            <sparkle:shortVersionString>1.2.1</sparkle:shortVersionString>
            <sparkle:minimumSystemVersion>10.13</sparkle:minimumSystemVersion>
            <enclosure url="https://github.com/kthchew/PolyMC/releases/download/sparkle-1/PolyMC.tar.gz" length="22843287" type="application/octet-stream" sparkle:edSignature="VjHIfxB/vb5gwGbqYB58NFOrsXk9Sd43b7Y5UC5MLTHB62Igco7ae47dth6IhvcQj0nMu/o2g3bTomfY49BODA=="/>
        </item>
    </channel>
</rss>
```

If a user were on a hypothetical version older than 1.2.1 that has this implemented, they would see this:
Screen Shot 2022-04-21 at 4 12 04 PM

(Yes, the release notes are ugly; it can render HTML, so ideally something not plain text would be used, but you can just change the RSS entry to point to a different URL with some nicer HTML to get the release notes. Technically you could even just point it to https://polymc.org/news/release-1.2.1/ but then the header on that site takes up a lot of space)

It both automatically checks for updates (at most, once every 24 hours, but we can change this if we want or give users an option to change it) and allows the user to manually check for updates. Before automatically checking for updates, it asks on second launch whether the user wants to automatically check or not.

---

### Stuff still in progress:

- [x] Config options for the updater in PolyMC settings are not functional (auto update settings ~and channel selection~)
- [x] Massive code cleanup needed; pretty much all of the old updater code is not necessary in macOS with this change, but I kind of just shoe horned it in. (kinda done?)
- [x] A lot of cmake warnings during install/packaging (still seems to work properly though). I'm not very good with cmake so I may need some help with this one.

---

### macOS is not required for releasing updates

I mentioned in #466 that macOS is needed to sign the update (to get the `edSignature` you see above). Actually, this isn't true. If you have a recent version of `openssl` (3.0 or higher, I think), you can just use the `openssl` command to create an ed25519 key:

```
openssl genpkey -algorithm ed25519 > priv.pem
openssl pkey -outform DER -pubout -in priv.pem | tail -c +13 | openssl base64 > pub.pem
```

and then sign the PolyMC archive (make sure to remove the newline from the output before putting it in the RSS entry):

```
openssl pkeyutl -sign -rawin -in /path/to/PolyMC.tar.gz -inkey priv.pem | openssl base64
```

so macOS is not necessary at all during the release process. It does seem to be necessary, though, if you want to create delta files (to allow for partial downloads), but this isn't required.

---

### Testing this PR

For testing purposes, the PR currently uses my public key and appcast URL, and places them in the `Info.plist`. If this is used, my key and URL must be removed before merging and changed to ones controlled by maintainers.

It also points to a "release" file on my personal fork. Unfortunately I can't directly point the appcast at the artifact generated by GitHub Actions, because GitHub Actions embeds the `.tar.gz` in a `.zip`, and then Sparkle doesn't know what to do with that.

Obviously, since there isn't a version above 1.2.1, if you download the artifact for this PR it will say there are no new updates available. You can trick it into thinking it's on an older version by changing the `CFBundleVersion` in the `PolyMC.app/Contents/Info.plist` to a lower version, then opening PolyMC again.

---

### Other operating systems

This PR does nothing for Windows or Linux. If this is used, I recommend using WinSparkle for Windows because the format of the `appcast.xml` is practically the same (in fact, you can even have the Mac/Windows versions point to the same appcast). However, it does seem like WinSparkle requires an installer to run (usually in some kind of unattended mode), so #66 needs to be implemented first. I'm not a Windows developer, though, so I'm not sure I can help too much with these.

From my understanding, Linux users can already use their package manager to update the launcher with the exception of AppImage users. I'm not too knowledgeable about that, but I don't think any update solutions there can pull from an appcast with a similar format.

### CODE REVIEW

**Comment by ZekeZDev** [comment]
The https://polymc.org/feed/feed.xml outputs html for each release I dont know if that can be used but I can probably intergrate something else for this pr if needed.

**Comment by kthchew** [comment]
> The https://polymc.org/feed/feed.xml outputs html for each release I dont know if that can be used but I can probably intergrate something else for this pr if needed.

I think that's actually perfect! If whatever produces that feed can be modified to add additional information to each entry, I think it can just use the same feed instead of a separate appcast. There's some documentation about how to create the appcast here: https://sparkle-project.org/documentation/publishing/#publishing-an-update

If you put HTML inside a `<description>` tag, Sparkle will just render that HTML directly (although it might need to be wrapped in `<![CDATA[ ... ]]>`). So yes, that works.

**Comment by DioEgizio** [comment]
I personally think that an updated should be implemented with an updater.txt or something like that and installed as a component (like portable builds).
This way if a packager wants to disable it they easily can

**Comment by DioEgizio** [comment]
after testing this, i ran that command but

> Domain (org.polymc.PolyMC) not found.
> Defaults have not been changed.

**Comment by kthchew** [comment]
> I personally think that an updated should be implemented with an updater.txt or something like that and installed as a component (like portable builds).
> This way if a packager wants to disable it they easily can

Currently, to disable the updater, you make the feed URL and/or the public key (both new build options) an empty string.

> after testing this, i ran that command but
> 
> > Domain (org.polymc.PolyMC) not found.
> > Defaults have not been changed.

Huh, that’s odd. Well, if that command doesn’t work, I guess there’s nothing to clear so it should be fine. There’s no file in `~/Library/Preferences/org.polymc.PolyMC.plist`?

Edit: Huh, apparently it doesn't save it there all the time. I just deleted my entire preferences for `org.polymc.PolyMC` and tried to test this PR again and `SUFeedURL` wasn't there. Weird. If it doesn't save it there, then no action is required.

Edit 2: Actually, I think that is a relic from one of my earlier times I tested it. I manually set the feed URL programmatically, which set that preference, but since I now use the `Info.plist` instead that won't appear. I'll remove that instruction from the main post.

**Comment by DioEgizio** [comment]
oh ok

**Comment by DioEgizio** [comment]
btw this looks really cool, thanks for doing this

**Comment by kthchew** [comment]
Okay, I'm gonna go ahead and mark this as ready for review. I still kept the old updater code around just in case, cause I'm not sure if it'll be used for something else or not. I tried to refactor stuff to make it easy to implement an updater on some other platform (just subclass the `ExternalUpdater` and implement the pure virtual functions). Some of those functions are unused (e.g. beta updates), but I added them in case they'll be used in the future.

Edit: cmake warnings/errors fixed

---

If someone in charge of releasing updates wants to test making a release, you'll need to first replace my public key and RSS URL with yours here:

https://github.com/PolyMC/PolyMC/blob/0dac6e4f48cdf0158dce580f1df024c7a4f2c828/CMakeLists.txt#L192-L193

Instructions for making the private/public key and signing is in my main post.

Note that if the "new" version does not contain an `SUPublicEDKey` in its `Info.plist`, then the update will fail. This means that you can't take the recent 1.2.1 release version and sign it, as it'll get rejected when you try to "update" to it. If you take a version that CI generated from this PR, it should be fine.

**Comment by kthchew** [comment]
> The https://polymc.org/feed/feed.xml outputs html for each release I dont know if that can be used but I can probably intergrate something else for this pr if needed.

@ZekeSmith regarding this, the below feed

```
<?xml version="1.0" standalone="yes"?>
<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" version="2.0">
    <channel>
        <title>PolyMC</title>
        <item>
            <title>1.2.1</title>
            <pubDate>Wed, 20 Apr 2022 23:45:13 -0400</pubDate>
            <sparkle:fullReleaseNotesLink>
                https://polymc.org/news/
            </sparkle:fullReleaseNotesLink>
            <description><![CDATA[
                <p>We have found some bugs in our last release on Sunday, which is why we are bringing you a truckload of bug fixes.<br /> Also check our friends' launch of the Quilt Beta over at <a href="https://quiltmc.org/">quiltmc.org</a>!</p> <h2 id="changelog" tabindex="-1">Changelog <a class="direct-link" href="https://polymc.org/news/release-1.2.1/#changelog" aria-hidden="true">#</a></h2> <h3 id="added" tabindex="-1">Added <a class="direct-link" href="https://polymc.org/news/release-1.2.1/#added" aria-hidden="true">#</a></h3> <ul> <li>Specify build platform for about page by @DioEgizio in #449</li> <li>[macOS] Enable hardened runtime by @kthchew in #454</li> </ul> <h3 id="changed" tabindex="-1">Changed <a class="direct-link" href="https://polymc.org/news/release-1.2.1/#changed" aria-hidden="true">#</a></h3> <h3 id="fixed" tabindex="-1">Fixed <a class="direct-link" href="https://polymc.org/news/release-1.2.1/#fixed" aria-hidden="true">#</a></h3> <ul> <li>Fix typos and capitalization by @kthchew in #442</li> <li>Fix placeholder translations for mod downloader by @flowln in #443</li> <li>Fix missing files in release workflow by @Scrumplex in #444</li> <li>Fix Modrinth search if Quilt is used by @Scrumplex in #450</li> <li>Fix crashes when creating a new instance by @Scrumplex in #451</li> <li>Fix crash when mod downloader has network issues by @Scrumplex in #452</li> <li>Fix sort-by-size in world list by @Scrumplex in #462</li> <li>Fix crash when trying to create an instance without a Minecraft version by @kthchew in #470</li> <li>Fix crash when opening the mod downloader with snapshots by @Scrumplex in #472</li> <li>[macOS] Fix version string by @kthchew in #455</li> <li>[macOS] Fix instance rename using return-key @txtsd in #463</li> </ul> <h3 id="removed" tabindex="-1">Removed <a class="direct-link" href="https://polymc.org/news/release-1.2.1/#removed" aria-hidden="true">#</a></h3> <ul> <li>[FreeBSD] Revert &quot;Force system LWJGL natives&quot; by @Scrumplex in #467</li> </ul> <p><strong>Full Changelog</strong>: <a href="https://github.com/PolyMC/PolyMC/compare/1.2.0...1.2.1">https://github.com/PolyMC/PolyMC/compare/1.2.0...1.2.1</a></p> <p>You can <a href="https://polymc.org/download">grab the latest download here</a> for your respective platform.</p>
                ]]>
            </description>
            <sparkle:version>1.2.1</sparkle:version>
            <sparkle:shortVersionString>1.2.1</sparkle:shortVersionString>
            <sparkle:minimumSystemVersion>10.13</sparkle:minimumSystemVersion>
            <enclosure url="https://github.com/kthchew/PolyMC/releases/download/sparkle-1/PolyMC.tar.gz" length="22843287" type="application/octet-stream" sparkle:edSignature="VjHIfxB/vb5gwGbqYB58NFOrsXk9Sd43b7Y5UC5MLTHB62Igco7ae47dth6IhvcQj0nMu/o2g3bTomfY49BODA=="/>
        </item>
    </channel>
</rss>
```

makes this much nicer update prompt
Screen Shot 2022-04-26 at 8 38 30 PM

So yeah, you should just be able to reuse that HTML

**Comment by ZekeZDev** [comment]
LGTM, what is needed on my end with website changes?

**Comment by kthchew** [comment]
@ZekeSmith The website should host an RSS feed ~(you mentioned it already has one for news, but you can also use a separate one if you'd like)~. Each entry related to an update should (at least) have the following tags:

```xml
<sparkle:version>$VERSION_NUMBER</sparkle:version>
<description><![CDATA[ $HTML_FOR_RELEASE_NOTES ]]>
<enclosure url="$LINK_TO_TAR_GZ" length="$SIZE_OF_TAR_GZ_IN_BYTES" type="application/octet-stream" sparkle:os="macos" sparkle:edSignature="$SIGNATURE_FOR_TAR_GZ"/>
<!--If WinSparkle is used for Windows, you can just add the below tags-->
<enclosure url="$LINK_TO_SETUP_EXE" length="$SIZE_OF_SETUP_EXE_IN_BYTES" type="application/octet-stream" sparkle:os="windows-x86" sparkle:dsaSignature="$SIGNATURE_FOR_SETUP_EXE"/>
<enclosure url="$LINK_TO_SETUP_EXE" length="$SIZE_OF_SETUP_EXE_IN_BYTES" type="application/octet-stream" sparkle:os="windows-x64" sparkle:dsaSignature="$SIGNATURE_FOR_SETUP_EXE"/>
```

Edit: `<rss xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle" version="2.0">` may also be necessary near the top of the feed

There's some other tags you can use if you'd like (for example, `<sparkle:minimumSystemVersion>10.13</sparkle:minimumSystemVersion>` is used to decide whether to offer an update to a user, e.g. if the minimum required macOS version is ever bumped and the user has a version below the new required version, they won't be asked to update). There's some information at https://sparkle-project.org/documentation/publishing/#publishing-an-update for other tags that may help in special cases, e.g. major updates, updates you want to be critical, etc.

To get the `SIGNATURE_FOR_TAR_GZ`, there's some instructions in my main post. I believe you'll need OpenSSL 3.0 or above. You could either do it manually for each release or automate it (the CI would need to securely keep the private key).

---

Once you've done that, there's two build options:

https://github.com/PolyMC/PolyMC/blob/5c220bbb526776b58d2b1b338031607d75918546/CMakeLists.txt#L192-L193

Replace my public key and URL with your public key used for the `SIGNATURE_FOR_TAR_GZ` and the RSS URL. Ideally, before merging you'd create a test release and then I (or someone else with a Mac) can verify that the release works. You should use one of the builds from this PR for the test release, because the update will fail if the “new” version has no public key in its `Info.plist`.

**Comment by DioEgizio** [comment]
This is conflicting with develop

**Comment by Scrumplex** [comment]
Is it possible to reduce the number of Sparkle files included in our repo? Like a Git submodule to upstream code?

**Comment by kthchew** [comment]
> Is it possible to reduce the number of Sparkle files included in our repo? Like a Git submodule to upstream code?

I think the only viable solution to that is to include the upstream code with a submodule, though that would require building Sparkle from source as well. Usually best practice is to use something like Swift Package Manager or CocoaPods instead to manage these dependencies, but since this isn't an Xcode project I don't think that's applicable.

**Comment by kthchew** [comment]
We could also have CMake download the Sparkle archive and then extract it (i.e. manually do what Swift Package Manager/CocoaPods would do).

**Comment by kthchew** [comment]
> We could also have CMake download the Sparkle archive and then extract it (i.e. manually do what Swift Package Manager/CocoaPods would do).

I did this ^ If this is acceptable, then I’ll see if I can clean up the Git history.

File: launcher/CMakeLists.txt
Comment (Scrumplex) [inline launcher/CMakeLists.txt]: Set these in the top-level CMakeLists file, preferably with the `CACHE` flag, so that these can be configured on build-time

**Comment by kthchew** [comment]
^ rebased with develop and cleaned up Git history. Now none of the commits have any Sparkle files in them.

File: CMakeLists.txt
Comment (Scrumplex) [inline CMakeLists.txt]: Is it maybe possible to somehow use GitHub releases instead?

Comment (kthchew) [reply CMakeLists.txt]: Do you mean to query for new versions instead of using an RSS feed? Unfortunately, I think it only uses appcasts. There’s actually an open Sparkle feature request for this very thing: https://github.com/sparkle-project/Sparkle/issues/648. The closest you can get to that right now is automatically generating the appcast when a new release is made.

Comment (Scrumplex) [reply CMakeLists.txt]: Hmm. We would need tooling for that. Maybe we can hack something with GitHub Actions and running a Pages instance for this repo.

Comment (kthchew) [reply CMakeLists.txt]: There's already a separate GitHub pages repo that contains an RSS feed for news, right? I remember @ZekeSmith saying it might be able to be adapted to implement this. If it's possible to add the [necessary tags](https://github.com/PolyMC/PolyMC/pull/479#issuecomment-1124516056) to that feed for release news, then it would probably work.

Comment (kthchew) [reply CMakeLists.txt]: Given that the CI uses a macOS machine to build the macOS version, it is also possible to use Sparkle's automatic tool to make the appcast (it's included in the release archive that I had CMake download, under `bin/generate_appcast`). Some stuff, like the download URL, may need to be edited after it generates the appcast, but that should be fairly easy.

Someone with a Mac would need to generate a private key using Sparkle's tool `bin/generate_keys` and add it to this repo's secrets though, because `openssl` seems to use a different private key format and I can't figure out how to migrate it from the OpenSSL format to the one the Sparkle tools accept.

**Comment by kthchew** [comment]
> If whatever produces that feed can be modified to add additional information to each entry, I think it can just use the same feed instead of a separate appcast.

I did some additional testing about my guess, and I think a separate feed would be needed. The feed at https://polymc.org/feed/feed.xml seems to be an Atom feed, but I think Sparkle only accepts RSS feeds. I tried putting the enclosure in the `<entry>`s but Sparkle was unable to detect an update.

**Comment by kthchew** [comment]
I made a PR over at the GitHub pages repo to generate the appcast. I think that the only manual thing left is signing the update (which could automatically generated with GitHub Actions) and adding the data to the release notes when making a new release (but I assume someone's already manually updating the release notes).

**Comment by kthchew** [comment]
Trying to get GHA to automatically output the signature for the archive as long as there is a `SPARKLE_ED25519_KEY` secret in the repo (formatted like a private key file). I don't think I can test it, because every time I do on my fork, the secret appears empty (even if I add a `SPARKLE_ED25519_KEY` secret on my own fork), unless I'm just doing something wrong. In theory, this should work if the secret isn't empty.

Edit: Nevermind. I tried to clone the repo into a private repo, and the secret is still empty. I'll try again later.

**Review by Scrumplex — APPROVED** [review]
Code looks fine

**Comment by Scrumplex** [comment]
Not sure if we will be able to get this setup and merged until tomorrow. Should we postpone this?

**Comment by kthchew** [comment]
Got the GHA signing working. If an administrator generates their private/public ed25519 keys with OpenSSL >=3.0:

```
openssl genpkey -algorithm ed25519 > priv.pem
openssl pkey -outform DER -pubout -in priv.pem | tail -c +13 | openssl base64 > pub.pem
```

...and adds a secret named `SPARKLE_ED25519_KEY` with the contents of the `priv.pem`, then GHA will automatically output the signature in the job summary ([example](https://github.com/kthchew/PolyMC/actions/runs/2363743703)).

> Not sure if we will be able to get this setup and merged until tomorrow. Should we postpone this?

No worries. Since there's another PR related to this in the website repo that would need to be reviewed before the release, if the new release is tomorrow then it probably makes sense to postpone it instead of rushing. Also, if this is added then macOS will have an updater but Windows won't, so maybe it makes more sense to add them both in the same release.

**Review by d-513 — APPROVED** [review]

**Review by ZekeZDev — APPROVED** [review]

**Comment by ZekeZDev** [comment]
are we good to merge this?

**Comment by Scrumplex** [comment]
We need to set up the server-side part

**Comment by kthchew** [comment]
> are we good to merge this?

Not yet. Here’s the remaining steps needed from a maintainer:

1. Install OpenSSL 3 or later on your local machine
2. Run the commands given in https://github.com/PolyMC/PolyMC/pull/479#issuecomment-1133692883 to generate the keys
3. Add a new secret to this repo named `SPARKLE_ED25519_KEY` with the contents of `priv.pem`. This will allow GHA to automatically generate the signatures. Make a backup of this key, because it’s required to release updates.
4. In `CMakeLists.txt`, set `MACOSX_SPARKLE_UPDATE_PUBLIC_KEY` to the contents of `pub.pem` and set `MACOSX_SPARKLE_UPDATE_FEED_URL` to `https://polymc.org/feed/appcast.xml`. (The current values are used for testing.)

https://github.com/PolyMC/PolyMC/blob/3bc02b9662b84c2ab86b5de1b08b4537177fde90/CMakeLists.txt#L198-L199

Changes to the release process are outlined in https://github.com/PolyMC/polymc.github.io/pull/139. Essentially, when you write the release notes on PolyMC.org, you need to add the required information at the top (alongside the title, tags, etc).

**Comment by BeansBeefBroccoli** [comment]
What is the blocked label?

**Comment by DioEgizio** [comment]
> What is the blocked label?

rn conflicting with develop, requires infrastructure changes

**Comment by Scrumplex** [comment]
I added a private key as a secret to GHA. The public key is `idALcUIazingvKSSsEa9U7coDVxZVx/ORpOEE/QtJfg=`

**Review by flowln — APPROVED** [review]
WRONG PR SORRY
LOOKING GREAT :+1:

**Comment by kthchew** [comment]
> WRONG PR SORRY LOOKING GREAT 👍

thank you for the support lol

> I added a private key as a secret to GHA. The public key is `idALcUIazingvKSSsEa9U7coDVxZVx/ORpOEE/QtJfg=`

Thanks, I'll add that once I get around to fixing the conflicts

**Comment by Scrumplex** [comment]
This PR should mostly resolve your conflicts: https://github.com/PolyMC/PolyMC/pull/876

**Comment by DioEgizio** [comment]
That pr has been merged

**Comment by Scrumplex** [comment]
@kthchew You commits need to be signed-off. See https://github.com/PolyMC/PolyMC/pull/479/checks?check_run_id=7272205108

**Comment by kthchew** [comment]
Will do! I’m not at my computer right now, but I should be able to get to it later today.

**Comment by kthchew** [comment]
This should now be ready for merging.

**Pull request merged**
